x/iro/keeper/settle.go

```diff
@@ -0,0 +1,150 @@
+package keeper
+
+import (
+	"errors"
+	"fmt"
+
+	errorsmod "cosmossdk.io/errors"
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	appparams "github.com/dymensionxyz/dymension/v3/app/params"
+	"github.com/dymensionxyz/dymension/v3/x/iro/types"
+	lockuptypes "github.com/dymensionxyz/dymension/v3/x/lockup/types"
+	"github.com/dymensionxyz/gerr-cosmos/gerrc"
+	balancer "github.com/osmosis-labs/osmosis/v15/x/gamm/pool-models/balancer"
+	gammtypes "github.com/osmosis-labs/osmosis/v15/x/gamm/types"
+)
+
+// AfterTransfersEnabled called by the genesis transfer IBC module when a transfer is handled
+// This is a rollapp module hook
+func (k Keeper) AfterTransfersEnabled(ctx sdk.Context, rollappId, rollappIBCDenom string) error {
+	return k.Settle(ctx, rollappId, rollappIBCDenom)
+}
+
+// Settle settles the iro plan with the given rollappId
+func (k Keeper) Settle(ctx sdk.Context, rollappId, rollappIBCDenom string) error {
+	plan, found := k.GetPlanByRollapp(ctx, rollappId)
+	if !found {
+		return nil
+	}
+
+	if plan.IsSettled() {
+		return errorsmod.Wrapf(errors.Join(gerrc.ErrInternal, types.ErrPlanSettled), "rollappId: %s", rollappId)
+	}
+
+	// validate the required funds are available in the module account
+	// funds expected as it's validated in the genesis transfer handler
+	balance := k.BK.GetBalance(ctx, k.AK.GetModuleAddress(types.ModuleName), rollappIBCDenom)
+	if !balance.Amount.Equal(plan.TotalAllocation.Amount) {
+		return errorsmod.Wrapf(gerrc.ErrInternal, "required: %s, available: %s", plan.TotalAllocation.String(), balance.String())
+	}
+
+	// "claims" the unsold FUT token
+	futBalance := k.BK.GetBalance(ctx, k.AK.GetModuleAddress(types.ModuleName), plan.TotalAllocation.Denom)
+	err := k.BK.BurnCoins(ctx, types.ModuleName, sdk.NewCoins(futBalance))
+	if err != nil {
+		return err
+	}
+
+	// mark the plan as `settled`, allowing users to claim tokens
+	plan.SettledDenom = rollappIBCDenom
+	k.SetPlan(ctx, plan)
+
+	// uses the raised DYM and unsold tokens to bootstrap the rollapp's liquidity pool
+	err = k.bootstrapLiquidityPool(ctx, plan)
+	if err != nil {
+		return errors.Join(types.ErrFailedBootstrapLiquidityPool, err)
+	}
+
+	// Emit event
+	err = ctx.EventManager().EmitTypedEvent(&types.EventSettle{
+		RollappId: rollappId,
+		PlanId:    fmt.Sprintf("%d", plan.Id),
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Bootstrap liquidity pool with the raised DYM and unsold tokens
+func (k Keeper) bootstrapLiquidityPool(ctx sdk.Context, plan types.Plan) error {
+	unallocatedTokens := plan.TotalAllocation.Amount.Sub(plan.SoldAmt)        // assumed > 0, as we enforce it in the Buy function
+	raisedDYM := k.BK.GetBalance(ctx, plan.GetAddress(), appparams.BaseDenom) // assumed > 0, as we enforce it by IRO creation fee
+
+	// send the raised DYM to the iro module
+	err := k.BK.SendCoinsFromAccountToModule(ctx, plan.GetAddress(), types.ModuleName, sdk.NewCoins(raisedDYM))
+	if err != nil {
```

Comment:
```suggestion
	// send the raised DYM to the iro module as it will be used as the pool creator
	err := k.BK.SendCoinsFromAccountToModule(ctx, plan.GetAddress(), types.ModuleName, sdk.NewCoins(raisedDYM))
```